pkg/pillar/localcommand/agent.go

```diff
@@ -0,0 +1,564 @@
+// The actual interval is chosen randomly between 0.8×1.1×interval (0.88×interval)
+// and 1.1×interval, which introduces jitter to avoid synchronized task execution
+// across multiple Go routines.
+func newTaskTicker(interval time.Duration) *taskTicker {
+	maxTime := 1.1 * float64(interval)
+	minTime := 0.8 * maxTime
+	t := &taskTicker{
+		tickerHandle: flextimer.NewRangeTicker(
+			time.Duration(minTime), time.Duration(maxTime)),
+	}
+	return t
+}
+
+// update sets the throttle state and recalculates the ticker's randomized interval.
+// When throttled is true, callers are expected to pass a longer interval (e.g. 1h
+// instead of 1m), and the ticker will continue to fire within the new randomized
+// [0.8×1.1×interval, 1.1×interval] range.
+func (t *taskTicker) update(throttle bool, interval time.Duration) {
+	t.Lock()
+	defer t.Unlock()
+	t.throttled = throttle
+	maxTime := 1.1 * float64(interval)
+	minTime := 0.8 * maxTime
+	t.tickerHandle.UpdateRangeTicker(time.Duration(minTime), time.Duration(maxTime))
+}
+
+// tickerIsThrottled reports whether the ticker is currently throttled.
+func (t *taskTicker) tickerIsThrottled() bool {
+	t.RLock()
+	defer t.RUnlock()
+	return t.throttled
+}
+
+// tickNow forces the ticker to fire immediately, unless throttled.
+func (t *taskTicker) tickNow() {
+	if t.tickerIsThrottled() {
+		return
+	}
+	t.tickerHandle.TickNow()
+}
+
+// tickerChan returns the ticker’s channel for receiving tick events.
+func (t *taskTicker) tickerChan() <-chan time.Time {
+	return t.tickerHandle.C
+}
+
+// taskControl provides thread-safe pause/resume functionality for concurrent tasks.
+// It allows multiple callers to pause task execution and requires all of them
+// to resume before tasks can continue running. Tasks can run in parallel when
+// not paused.
+type taskControl struct {
+	beforeStart  func()
+	pauseCounter int
+	pauseMx      sync.Mutex
+	taskMx       sync.RWMutex
+}
+
+// pause increments the pause counter and blocks new tasks from starting.
+// Multiple callers can pause concurrently - all must call resume before
+// tasks can continue. Returns a resume function that must be called to
+// decrement the pause counter.
+func (lc *taskControl) pause() (resume func()) {
+	lc.pauseMx.Lock()
+	defer lc.pauseMx.Unlock()
+
+	lc.pauseCounter++
+	if lc.pauseCounter == 1 {
+		// Pause begins here.
+		lc.taskMx.Lock()
+	} // else already paused
+
+	return func() {
+		lc.pauseMx.Lock()
+		defer lc.pauseMx.Unlock()
+
+		lc.pauseCounter--
+		if lc.pauseCounter == 0 {
+			// Pause ends here.
+			lc.taskMx.Unlock()
+		}
+	}
+}
+
+// startTask marks the beginning of a task.
+//   - If a beforeStart callback is configured, it is executed before attempting to start.
+//   - Returns true if task execution is currently paused, in which case the caller
+//     should not proceed.
+func (lc *taskControl) startTask() (paused bool) {
+	if lc.beforeStart != nil {
+		lc.beforeStart()
+	}
+	locked := lc.taskMx.TryRLock()
```

Comment:
Did we hold a (read) lock across http invocations in the old/current implementation?
Means the read lock can be held for minutes.

Reply:
We didn't (but then had some potential race conditions in the old implementation).
But I changed this now based on your comments to use `context.ContextWithCancel` for LPS HTTP operations, and this context will get canceled on Pause.